### PR TITLE
Support nested configurations

### PR DESCRIPTION
**Pull request by tor-oscar**
Add support for nested configurations to allow fetching values in a hierarchical structure.

Adds support for Map.fetch in addition to Application.fetch_env and Keyword.fetch.

### CODE REVIEW

File: lib/app_config.ex [hunk 1]
```diff
@@ -132,21 +133,41 @@ defmodule AppConfig do
...
   """
-  @spec fetch_env(app | Keyword.t, key) :: {:ok, value} | :error
-  def fetch_env(env, key) when (is_atom(env) or is_list(env)) and is_atom(key) do
+  @spec fetch_env(app | Keyword.t, key | List.t) :: {:ok, value} | :error
+  def fetch_env(env, key) when (is_atom(env) or is_list(env) or is_map(env)) and is_atom(key) do
     with {:ok, value} <- fetch(env, key) do
       get_env_value(value)
     end
   end
+  def fetch_env(env, [ first_key | keys ]) when is_atom(env) or is_list(env) or is_map(env) do
```
Comment (jcomellas) [inline lib/app_config.ex]: @DrunkenInfant Could you please keep the spacing in brackets and braces consistent with the rest of the code? In this case, the expression should be `[first_key | keys]`.

**Review by jcomellas — CHANGES_REQUESTED** [review]
Thanks for this feature. Sorry for the radio silence during 15 days, but the GitHub notification for this project got lost in the morass of notifications from other projects.

I left just a few minor comments. Also, could you please rebase before pushing the new changes?

File: lib/app_config.ex [hunk 2]
```diff
@@ -132,21 +133,41 @@ defmodule AppConfig do
...
+
+  defp fetch_nested(key, { :ok, val }) do
+    case val do
+      val when is_map(val) or is_list(val) -> fetch(val, key)
+      _ -> :error
+    end
+  end
+  defp fetch_nested(_, :error) do
+    :error
+  end
```
Comment (jcomellas) [inline lib/app_config.ex]: The `fetch_nested/2` function could be simplified a bit. Maybe to something like:
```elixir
defp fetch_nested(key, {:ok, val}) when is_map(val) or is_list(val) do
  fetch(val, key)
end
defp fetch_nested(_key, _result) do
  :error
end
```

**Comment by tor-oscar** [comment]
Sorry for the delay. I fixed your comments.

**Pull request closed without merging**